components/org.wso2.carbon.metrics.view.ui/src/main/resources/web/metrics-view/metrics_hierarchy_menu_data_ajaxprocessor.jsp

```diff
@@ -0,0 +1,176 @@
+<%@ page import="org.wso2.carbon.CarbonConstants" %>
+<%@ page import="org.wso2.carbon.metrics.data.common.*" %>
+<%@ page import="org.wso2.carbon.metrics.view.ui.MetricDataWrapper" %>
+<%@ page import="org.wso2.carbon.metrics.view.ui.MetricHierarchyDataWrapper" %>
+<%@ page import="org.wso2.carbon.metrics.view.ui.MetricMetaWrapper" %>
+<%@ page import="org.wso2.carbon.metrics.view.ui.MetricsViewClient" %>
+<%@ page import="org.wso2.carbon.ui.CarbonUIUtil" %>
+<%@ page import="org.wso2.carbon.utils.ServerConstants" %>
+<%@ page import="java.util.ArrayList" %>
+<%@ page import="java.util.List" %>
+
+<%
+    String source = request.getParameter("source");
+    String from = request.getParameter("from");
+    String to = request.getParameter("to");
+    String path = request.getParameter("path");
+    String type = request.getParameter("type");
+    path = (path != null && path.trim().length() > 0) ? path : "";
+    type = (type != null && type.trim().length() > 0) ? type : "";
+
+    String backendServerURL = CarbonUIUtil.getServerURL(config.getServletContext(), session);
+    ConfigurationContext configContext = (ConfigurationContext) config.getServletContext().getAttribute(
+            CarbonConstants.CONFIGURATION_CONTEXT);
+    String cookie = (String) session.getAttribute(ServerConstants.ADMIN_SERVICE_COOKIE);
+    MetricsViewClient metricsViewClient;
+    try {
+        metricsViewClient = new MetricsViewClient(cookie, backendServerURL, configContext);
+        Gson gson = new Gson();
+        MetricDataWrapper metricData = null;
+        ArrayList<Metric> metrics = getMetrics(metricsViewClient, source, path, type);
+        MetricList metricList = new MetricList();
+        metricList.setMetric(metrics.toArray(new Metric[metrics.size()]));
+        if (to != null && to.trim().length() > 0) {
+            metricData = metricsViewClient.findMetricsByTimePeriod(metricList, source,
+                    Long.parseLong(from), Long.parseLong(to));
+        } else {
+            metricData = metricsViewClient.findLastMetrics(metricList, source, from);
+        }
+        if (metricData != null) {
+            response.getWriter().write(gson.toJson(metricData));
+        }
+    } catch (Exception e) {
+        return;
+    }
+    response.setContentType("application/json");
+%>
+
+<%!
+    private MetricHierarchyDataWrapper getHierarchyData(MetricsViewClient client, String source, String path) {
+        MetricHierarchyDataWrapper metricHierarchy;
+        try {
+            metricHierarchy = client.getHierarchy(source, path);
+        } catch (Exception e) {
+            metricHierarchy = null;
+        }
+        return metricHierarchy;
+    }
+%>
+
+<%!
+    private ArrayList<Metric> getMetrics(MetricsViewClient client, String source, String path, String type) {
+        MetricHierarchyDataWrapper hierarchyData = getHierarchyData(client, source, path);
+        ArrayList<Metric> metrics = new ArrayList<Metric>();
```

Comment:
Always make return types bound to appropriate abstractions for better extensibility of the code. For instance, in this particular case, you could have used "List<Metric>" as the return type of the method, which will let you change the underlying List implementation seamlessly to something else other than an ArrayList, should that be required.